eval/adjust.rkt

```diff
@@ -71,333 +71,263 @@
-     (define x (first srcs))
-     (define y (second srcs))
-     (define ylo (ival-lo y))
-     (define yhi (ival-hi y))
-     (define outlo (ival-lo output))
-     (define outhi (ival-hi output))
-     
-     (define x-exp (ival-max-log2-approx x))
-     (define out-exp (ival-min-log2-approx output))
-     
-     (define x-slack (if (equal? (mpfr-sign outlo) (mpfr-sign outhi))
-                         0                
-                         (get-slack)))                ; output crosses 0
-     
-     (define y-slack (if (equal? (mpfr-sign ylo) (mpfr-sign yhi))
-                         x-slack                
-                         (+ x-slack (get-slack))))    ; y crosses zero
-     
-     (list (+ (- x-exp out-exp) 1 x-slack)            ; exponent per x
-           (+ (- x-exp out-exp) 1 y-slack))]          ; exponent per y
-    
-    [(ival-fma)
-     ; log[Гfma] = log[ max(x*y, -z) / fma(x,y,z)] ~ max(log[x] + log[y], log[z]) - log[fma(x,y,z)] + 1
-     ;                               ^^^^^^^^^^^^
-     ;                               possible uncertainty
-     (define x (first srcs))
-     (define y (second srcs))
-     (define z (third srcs))
-     (define outlo (ival-lo output))
-     (define outhi (ival-hi output))
-     
-     (define x-exp (ival-max-log2-approx x))
-     (define y-exp (ival-max-log2-approx y))
-     (define z-exp (ival-max-log2-approx z))
-     (define out-exp (ival-min-log2-approx output))
-     
-     (define slack (if (equal? (mpfr-sign outhi) (mpfr-sign outlo))
-                                0
-                                (get-slack)))         ; cancellation when output crosses 0
-     
-     (define lhs-exp (max (+ x-exp y-exp)             ; max(log[x] + log[y], log[z])
-                          z-exp))
-     (make-list 3 (+ (- lhs-exp out-exp) 1 slack))]   ; exponents per arguments
-    
-    [(ival-hypot)
-     ; hypot = sqrt(x^2+y^2)
-     ; log[Гhypot] = log[ (2 * max(x,y) / hypot(x,y))^2 ] = 2 * (1 + log[max(x,y)] - log[hypot]) + 1
-     ;                                  ^
-     ;                                  a possible division by zero, catched by log2-approx's slack
+     ; k = 1: maxlog(x) - minlog(z)
+     ; k = 2: ~ log[y * rnd_down(x/y)] - log[mod(x,y)] <= maxlog(x) - minlog(z)
+     ;                            ^    ^
+     ;                           conditions of uncertainty
      (define x (first srcs))
      (define y (second srcs))
      
-     (define x-exp (ival-max-log2-approx x))
-     (define y-exp (ival-max-log2-approx y))
-     (define out-exp (ival-min-log2-approx output))
+     (define slack (if (crosses-zero? y)
+                       0                
+                       (get-slack)))                  ; y crosses zero
      
-     (make-list 2 (+ (* 2 (- (+ 1 (max x-exp y-exp)) out-exp)) 1))]
+     (list (- (maxlog x) (minlog z))                  ; exponent per x
+           (+ (- (maxlog x) (minlog z)) slack))]      ; exponent per y
```

Comment:
Ok whatever, I don't think we are super sure of the formulas for `fmod` but who cares.

Reply:
Yeah, I am also not sure about `slack` here, I suppose it does not make it worse. But maybe it should be removed